File: js_tools/list_directory.js
return function(args) {
  args = args || {};
  const path = args.path || ".";

  function parseOptionalInteger(value, fallback) {
    if (value === undefined || value === null) return fallback;
    if (typeof value === "number" && Number.isInteger(value)) return value;
    if (typeof value === "string" && /^-?\d+$/.test(value.trim())) {
      return Number.parseInt(value, 10);
    }
    return fallback;
  }

  const rawDepth = parseOptionalInteger(args.depth, 1);
  const depth = Math.max(1, Math.min(8, rawDepth));

  const includeIgnored = args.include_ignored === true;
  if (args.ignore !== undefined && !Array.isArray(args.ignore)) {
    throw new Error("INVALID_ARGUMENT: ignore must be an array of strings");
  }

  const defaultIgnores = [".git", "node_modules", "bazel-*", "dist", "build", ".cache", ".next", "target"];
  const customIgnore = Array.isArray(args.ignore)
    ? args.ignore.filter(p => typeof p === "string" && p.trim() !== "")
    : [];

  const ignorePatterns = includeIgnored ? [] : defaultIgnores.concat(customIgnore);

  let pruneClause = "";
  if (ignorePatterns.length > 0) {
    pruneClause = " \\( " + ignorePatterns.map(p => "-name " + shell_escape(p)).join(" -o ") + " \\) -prune -o";
  }

  const cmd =
    "find " + shell_escape(path) +
    " -mindepth 1 -maxdepth " + depth +
    pruneClause +
    " -mindepth 1 -maxdepth " + depth +
    " -printf '%y\t%P\n'";

  const res = __os_run(cmd);
  if (res.exit_code !== 0) {
    throw new Error("Failed to list directory: " + res.stderr);
  }

  const lines = res.stdout.split("\n").filter(l => l.trim() !== "");
  const output = [];

  for (const line of lines) {
    const tabIdx = line.indexOf("\t");
    if (tabIdx === -1) continue;

    const type = line.substring(0, tabIdx);
    const rel = line.substring(tabIdx + 1);
    if (!rel) continue;

    if (type === "d") {
      output.push("Directory: " + rel + "/");
    } else {
      output.push("File: " + rel);
    }
  }

  return output.join("\n");
};

File: tools/list_directory.cpp
#include <sstream>
#include <vector>

#include "absl/status/status.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/str_join.h"

#include "core/json_utils.h"
#include "core/shell_util.h"
#include "core/status_macros.h"
#include "tools/tool_executor.h"

namespace slop {
absl::StatusOr<std::string> ToolExecutor::HandleListDirectory(const nlohmann::json& args) {
  const std::string path = json_get_or<std::string>(args, "path", ".");

  int raw_depth = 1;
  if (auto depth_int = json_get<int>(args, "depth")) {
    raw_depth = *depth_int;
  } else if (auto depth_str = json_get<std::string>(args, "depth")) {
    if (!absl::SimpleAtoi(*depth_str, &raw_depth)) {
      return absl::InvalidArgumentError("depth must be an integer");
    }
  }
  const int depth = std::max(1, std::min(8, raw_depth));

  const bool include_ignored = json_get_or<bool>(args, "include_ignored", false);
  std::vector<std::string> ignore_patterns;
  if (!include_ignored) {
    ignore_patterns = {".git", "node_modules", "bazel-*", "dist", "build", ".cache", ".next", "target"};
    const auto* ignore_json = json_at(args, "ignore");
    if (ignore_json && !ignore_json->is_null()) {
      if (!ignore_json->is_array()) {
        return absl::InvalidArgumentError("ignore must be an array of strings");
      }
      for (const auto& it : *ignore_json) {
        if (it.is_string() && !it.get<std::string>().empty()) {
          ignore_patterns.push_back(it.get<std::string>());
        }
      }
    }
  }

  std::string prune_clause;
  if (!ignore_patterns.empty()) {
    std::vector<std::string> names;
    names.reserve(ignore_patterns.size());
    for (const auto& p : ignore_patterns) {
      names.push_back(absl::StrCat("-name ", EscapeShellArg(p)));
    }
    prune_clause = absl::StrCat(" \\( ", absl::StrJoin(names, " -o "), " \\) -prune -o");
  }

  const std::string cmd = absl::StrCat(
      "cd ", EscapeShellArg(path), " && find . ", " -mindepth 1 -maxdepth ", depth, prune_clause,
      " -mindepth 1 -maxdepth ", depth,
      " -exec sh -c 'for f; do rel=\"${f#./}\"; if [ -d \"$f\" ]; then printf \"d\\t%s\\n\" \"$rel\"; else "
      "printf \"f\\t%s\\n\" \"$rel\"; fi; done' sh {} +");

  ASSIGN_OR_RETURN(auto run_res, RunCommand(cmd));
  if (run_res.exit_code != 0) {
    return absl::InternalError(absl::StrCat("Failed to list directory: ", run_res.stderr_out));
  }

  std::vector<std::string> out;
  std::stringstream ss(run_res.stdout_out);
  std::string row;
  while (std::getline(ss, row)) {
    if (row.empty()) continue;
    size_t tab = row.find('\t');
    if (tab == std::string::npos) continue;
    std::string type = row.substr(0, tab);
    std::string rel = row.substr(tab + 1);
    if (rel.empty()) continue;
    if (type == "d") {
      out.push_back(absl::StrCat("Directory: ", rel, "/"));
    } else {
      out.push_back(absl::StrCat("File: ", rel));
    }
  }
  return absl::StrJoin(out, "\n");
}

}  // namespace slop
